<a href="https://colab.research.google.com/github/Narawit007/BSC_DPDM2025/blob/main/Midterm_663020283_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# google drive

https://drive.google.com/drive/folders/10n2BtWfV_NkT57HB-WzP_KxEQFJefzwe?usp=sharing

#ตอบคำถาม

- ความหมายแต่ละคอลัม

Node - สถานี

rain - ปริมาณน้ำฝน

month - เดือน(เป็นเลข)

year - ปี

---------------------------
rainfall_datetime - เวลาที่อ้างอิงของข้อมูลฝน

rainfall_datetime_calc - เวลาที่ปรับจากหลายๆที่

rainfall_value - ปริมาณฝน

tele_station_id - รหัสสถานี(ใช้ในที่ทำงาน)

tele_station_oldcode - ชื่อหรือรหัสสถานี(ที่ระบุได้จริง)

measure_datetime - วันที่วัด


----------------------------------

- เหตุผลที่เรานิยมใช้ค่า Median แทน Mean
1. ข้อมูลน้ำฝนมีการกระจายตัวแบบไม่ปกติ
2. ค่าที่โดดไปจากเพื่อน
3. ข้อมูลส่วนใหญ่เป็นค่าศูนย์
```

```


#ดึงข้อมูลเข้ามา

In [20]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [21]:
import pandas as pd

file_path = '/content/drive/MyDrive/สอบ3เทอม2/Input - (Operation Daily rain from API_ 2025)/Stations in the model_HII.csv'

try:
    df_stations = pd.read_csv(file_path)
    if 'Node' in df_stations.columns:
        unique_nodes = df_stations['Node'].unique()
        print("รายชื่อ Node ทั้งหมดในไฟล์ Stations in the model_HII.csv:")
        for node in unique_nodes:
            print(node)
    else:
        print(f"ไม่พบคอลัมน์ 'Node' ในไฟล์ {file_path}")
except FileNotFoundError:
    print(f"ไม่พบไฟล์ที่ระบุ: {file_path}")
except Exception as e:
    print(f"เกิดข้อผิดพลาดในการอ่านไฟล์หรือประมวลผลข้อมูล: {e}")

    num_unique_nodes = len(unique_nodes)
print(f"จำนวน Node ทั้งหมดที่พบในไฟล์ Stations in the model_HII.csv คือ: {num_unique_nodes}")

รายชื่อ Node ทั้งหมดในไฟล์ Stations in the model_HII.csv:
ACRU
BBHN
BBUA
BCNG
BDAR
BDCP
BDGN
BDLH
BDMG
BGRT
BHMN
BHRA
BHUN
BJIG
BKDN
BKHL
BKHN
BKK001
BKK008
BKK009
BKK015
BKK017
BKK018
BKK019
BKUG
BKWN
BLAT
BLKO
BLUG
BMDG
BMNK
BNAN
BNGR
BNHG
BNHI
BNHO
BNKN
BNKP
BNLG
BNMK
BNPI
BNPU
BNTK
BOKA
BPIA
BPK004
BPLA
BPPS
BPTG
BRKM
BSBA
BSDE
BSKJ
BSMP
BSPN
BSUM
BTHO
BWKG
BYNU
CATK
CBUR
CCSS
CEHM
CGDO
CGKM
CGMN
CGSN
CHHA
CHI005
CHI010
CHI011
CHI012
CHI013
CHM001
CHM002
CHM003
CHM004
CHN004
CHNA
CHPA
CHR003
CHR004
CHR005
CHUN
CLPK
CMSG
CPKC
CPY003
CPY004
CPY009
CPY010
CPY012
CPY013
CPY015
CPY017
CSMO
CTKN
DCCT
DGSL
DITO
DIV001
DIV002
DIV003
DIV004
DIV005
DKLG
DKTI
DMKH
DNCM
DSKT
DTAN
HCTT
HKTG
HNKA
HNRE
HSYO
HTTM
HUYG
HYOD
HYRT
JBLG
KABG
KAHN
KAKG
KANG
KAOP
KGKG
KGLN
KHHM
KHPB
KKHA
KKPA
KKTO
KKWF
KMSI
KNYM
KOKN
KOLO
KON002
KON004
KON005
KOSB
KOSK
KOSR
KPEM
KRKM
KRN002
KSOD
KSPS
KTKB
KTLE
KWSI
LBI002
LEEE
LK13
LMPN
LMSK
LNRG
LPLE
MCHI
MCRM
MDTG
MECN
MEPK
MEWG
MKTR
MMMO
MMNY
MNRM
MRMT
MUN011
MUN014


In [22]:
num_unique_nodes = len(unique_nodes)
print(f"จำนวน Node ทั้งหมดที่พบในไฟล์ Stations in the model_HII.csv คือ: {num_unique_nodes}")

จำนวน Node ทั้งหมดที่พบในไฟล์ Stations in the model_HII.csv คือ: 385


#ขั้นตอนที่1 กรองเฉพาะสถานีที่อยู่ใน Stations in the model_HII.csv

(385สถานี)

In [23]:
import pandas as pd
import os

# --- ส่วนโหลดไฟล์ Master (ใช้ valid_nodes เดิม) ---
# (สมมติว่าคุณรัน valid_nodes ไว้แล้วจากโค้ดก่อนหน้า)

if valid_nodes is not None:
    print(f"{'Month':<15} | {'Avg Before':<12} | {'Avg After':<12} | {'Files Processed'}")
    print("-" * 60)

    for monthly_dir_path in monthly_directories:
        if not os.path.isdir(monthly_dir_path):
            continue

        folder_name = os.path.basename(monthly_dir_path)
        csv_files = [f for f in os.listdir(monthly_dir_path) if f.endswith('.csv')]

        total_before = 0
        total_after = 0
        count_files = 0

        for filename in csv_files:
            file_path = os.path.join(monthly_dir_path, filename)
            try:
                df_daily = pd.read_csv(file_path)

                # เช็คว่ามีคอลัมน์ที่ต้องการไหม
                if 'tele_station_oldcode' in df_daily.columns:
                    df_daily['tele_station_oldcode'] = df_daily['tele_station_oldcode'].astype(str)

                    stations_before = df_daily['tele_station_oldcode'].nunique()
                    df_filtered = df_daily[df_daily['tele_station_oldcode'].isin(valid_nodes)]
                    stations_after = df_filtered['tele_station_oldcode'].nunique()

                    total_before += stations_before
                    total_after += stations_after
                    count_files += 1

                    # หากต้องการเซฟทับไฟล์จริง ให้เอาเครื่องหมาย # ออกด้านล่างนี้:
                    # df_filtered.to_csv(file_path, index=False)

            except Exception:
                continue # ข้ามไฟล์ที่มีปัญหา

        # คำนวณค่าเฉลี่ยต่อเดือนเพื่อสรุป
        if count_files > 0:
            avg_before = int(total_before / count_files)
            avg_after = int(total_after / count_files)
            print(f"{folder_name:<15} | {avg_before:<12} | {avg_after:<12} | {count_files} files")
        else:
            print(f"{folder_name:<15} | No valid CSV files found.")

Month           | Avg Before   | Avg After    | Files Processed
------------------------------------------------------------
2025-07         | 886          | 364          | 31 files
2025-08         | 883          | 354          | 31 files
2025-09         | 886          | 349          | 30 files
2025-10         | 882          | 354          | 31 files
2025-11         | 964          | 370          | 30 files
2025-12         | 982          | 368          | 31 files


In [24]:
import pandas as pd
import os

# Define the base output directory for filtered files
base_output_dir = '/content/drive/MyDrive/สอบ3เทอม2/Output - Filtered Daily Rain'

# Ensure valid_nodes is available from unique_nodes
if 'unique_nodes' not in locals() or unique_nodes is None:
    print("Error: 'unique_nodes' ไม่พบ. โปรดรันเซลล์ก่อนหน้าเพื่อโหลด Node จากไฟล์ Master.")
    # Fallback in case unique_nodes was somehow not set, though it should be.
    try:
        master_file_path = '/content/drive/MyDrive/สอบ3เทอม2/Input - (Operation Daily rain from API_ 2025)/Stations in the model_HII.csv'
        df_master_stations = pd.read_csv(master_file_path)
        if 'Node' in df_master_stations.columns:
            valid_nodes = set(df_master_stations['Node'].unique())
            print("valid_nodes ถูกโหลดจากไฟล์ Master แล้ว (เป็น fallback).")
        else:
            raise ValueError(f"คอลัมน์ 'Node' ไม่พบในไฟล์ Master: {master_file_path}")
    except Exception as e:
        print(f"Error: เกิดข้อผิดพลาดในการโหลด valid_nodes (fallback): {e}")
        valid_nodes = set()
else:
    valid_nodes = set(unique_nodes) # Use the existing unique_nodes

# Ensure monthly_directories is available
if 'monthly_directories' not in locals():
    print("Error: 'monthly_directories' ไม่พบ. โปรดรันเซลล์ก่อนหน้าเพื่อกำหนดไดเรกทอรีรายเดือน.")
    # Fallback to define monthly_directories if not found
    base_input_dir = '/content/drive/MyDrive/สอบ3เทอม2/Input - (Operation Daily rain from API_ 2025)'
    monthly_directories = [
        os.path.join(base_input_dir, '2025-07'),
        os.path.join(base_input_dir, '2025-08'),
        os.path.join(base_input_dir, '2025-09'),
        os.path.join(base_input_dir, '2025-10'),
        os.path.join(base_input_dir, '2025-11'),
        os.path.join(base_input_dir, '2025-12')
    ]
    print("monthly_directories ถูกกำหนดใหม่แล้ว (เป็น fallback)." + "\n")

station_code_column = 'tele_station_oldcode'

print("กำลังประมวลผลและสร้างไฟล์ใหม่สำหรับข้อมูลรายวันที่กรองแล้ว...")
print(f"ไฟล์ที่กรองแล้วจะถูกบันทึกใน '{base_output_dir}'")
print("-" * 80)

total_files_processed = 0
total_stations_before_filter_sum = 0 # Sum of unique stations across all files before filter
total_stations_after_filter_sum = 0  # Sum of unique stations across all files after filter

for monthly_dir_path in monthly_directories:
    folder_name = os.path.basename(monthly_dir_path)
    output_monthly_dir = os.path.join(base_output_dir, folder_name)
    os.makedirs(output_monthly_dir, exist_ok=True) # Create output directory if it doesn't exist

    if not os.path.isdir(monthly_dir_path):
        print(f"  Warning: {monthly_dir_path} ไม่ใช่ไดเรกทอรีที่ถูกต้อง ข้ามไป.")
        continue

    csv_files = [f for f in os.listdir(monthly_dir_path) if f.endswith('.csv')]

    if not csv_files:
        print(f"  ไม่มีไฟล์ CSV พบใน {folder_name}. ข้ามไป.")
        continue

    files_in_month_processed = 0
    stations_before_month_avg_accumulator = []
    stations_after_month_avg_accumulator = []

    for filename in csv_files:
        file_path = os.path.join(monthly_dir_path, filename)
        output_file_path = os.path.join(output_monthly_dir, filename)

        try:
            df_daily = pd.read_csv(file_path)

            if station_code_column not in df_daily.columns:
                print(f"  Warning: คอลัมน์ '{station_code_column}' ไม่พบในไฟล์ {filename}. ข้ามไฟล์นี้.")
                continue

            df_daily[station_code_column] = df_daily[station_code_column].astype(str)

            # Count unique stations before filtering
            stations_before_this_file = df_daily[station_code_column].nunique()

            # Filter the DataFrame
            df_filtered = df_daily[df_daily[station_code_column].isin(valid_nodes)].copy()

            # Count unique stations after filtering
            stations_after_this_file = df_filtered[station_code_column].nunique()

            # Save the filtered DataFrame to the new output path
            df_filtered.to_csv(output_file_path, index=False)

            stations_before_month_avg_accumulator.append(stations_before_this_file)
            stations_after_month_avg_accumulator.append(stations_after_this_file)
            files_in_month_processed += 1
            total_files_processed += 1

        except FileNotFoundError:
            print(f"  Error: ไม่พบไฟล์ที่ระบุ: {file_path}. ข้ามไป.")
        except Exception as e:
            print(f"  Error: เกิดข้อผิดพลาดในการประมวลผลไฟล์ {file_path}: {e}. ข้ามไป.")

    if files_in_month_processed > 0:
        avg_before = int(sum(stations_before_month_avg_accumulator) / files_in_month_processed)
        avg_after = int(sum(stations_after_month_avg_accumulator) / files_in_month_processed)
        print(f"เดือน {folder_name}: เฉลี่ยสถานีก่อนกรอง: {avg_before}, เฉลี่ยสถานีหลังกรอง: {avg_after} ({files_in_month_processed} ไฟล์ที่ถูกประมวลผล)")
        total_stations_before_filter_sum += sum(stations_before_month_avg_accumulator)
        total_stations_after_filter_sum += sum(stations_after_month_avg_accumulator)
    else:
        print(f"เดือน {folder_name}: ไม่พบไฟล์ที่ประมวลผลได้.")

print("-" * 80)
print(f"เสร็จสิ้นการดำเนินการ. สร้างไฟล์ใหม่ที่กรองแล้วทั้งหมด {total_files_processed} ไฟล์ใน '{base_output_dir}'")
if total_files_processed > 0:
    print(f"ค่าเฉลี่ยรวมของสถานีก่อนกรองต่อวัน: {total_stations_before_filter_sum / total_files_processed:.0f}")
    print(f"ค่าเฉลี่ยรวมของสถานีหลังกรองต่อวัน: {total_stations_after_filter_sum / total_files_processed:.0f}")
else:
    print("ไม่มีไฟล์ใดถูกประมวลผล.")

กำลังประมวลผลและสร้างไฟล์ใหม่สำหรับข้อมูลรายวันที่กรองแล้ว...
ไฟล์ที่กรองแล้วจะถูกบันทึกใน '/content/drive/MyDrive/สอบ3เทอม2/Output - Filtered Daily Rain'
--------------------------------------------------------------------------------
เดือน 2025-07: เฉลี่ยสถานีก่อนกรอง: 886, เฉลี่ยสถานีหลังกรอง: 364 (31 ไฟล์ที่ถูกประมวลผล)
เดือน 2025-08: เฉลี่ยสถานีก่อนกรอง: 883, เฉลี่ยสถานีหลังกรอง: 354 (31 ไฟล์ที่ถูกประมวลผล)
เดือน 2025-09: เฉลี่ยสถานีก่อนกรอง: 886, เฉลี่ยสถานีหลังกรอง: 349 (30 ไฟล์ที่ถูกประมวลผล)
เดือน 2025-10: เฉลี่ยสถานีก่อนกรอง: 882, เฉลี่ยสถานีหลังกรอง: 354 (31 ไฟล์ที่ถูกประมวลผล)
เดือน 2025-11: เฉลี่ยสถานีก่อนกรอง: 964, เฉลี่ยสถานีหลังกรอง: 370 (30 ไฟล์ที่ถูกประมวลผล)
เดือน 2025-12: เฉลี่ยสถานีก่อนกรอง: 982, เฉลี่ยสถานีหลังกรอง: 368 (31 ไฟล์ที่ถูกประมวลผล)
--------------------------------------------------------------------------------
เสร็จสิ้นการดำเนินการ. สร้างไฟล์ใหม่ที่กรองแล้วทั้งหมด 184 ไฟล์ใน '/content/drive/MyDrive/สอบ3เทอม2/Output - Filtered Daily Rain'
ค่าเฉลี่ยรวมข

#ขั้นตอนที่2 ลบสถานีที่มีข้อมูลหาย >= 20%

In [25]:
import pandas as pd
import os
from calendar import monthrange

# --- ตั้งค่าเส้นทางโฟลเดอร์ที่เป็นผลลัพธ์การกรองแล้ว ---
output_filtered_path = '/content/drive/MyDrive/สอบ3เทอม2/Output - Filtered Daily Rain'
target_col = 'rainfall_value'
station_col = 'tele_station_oldcode'

# รายชื่อโฟลเดอร์เดือน
months = ['2025-07', '2025-08', '2025-09', '2025-10', '2025-11', '2025-12']

for month_str in months:
    dir_path = os.path.join(output_filtered_path, month_str)
    if not os.path.isdir(dir_path):
        continue

    print(f"\n📊 [ สรุปข้อมูลสูญหายรายสถานี: {month_str} ]")
    print(f"{'Station ID':<15} | {'Expected':<10} | {'Missing':<10} | {'Missing %'}")
    print("-" * 55)

    # 1. หาจำนวนวันทั้งหมดในเดือนนั้น (เพื่อรู้ว่า 1 สถานีควรมีกี่ Record)
    year, month = map(int, month_str.split('-'))
    days_in_month = monthrange(year, month)[1]

    # 2. รวมข้อมูลทุกไฟล์ในเดือนนั้นเข้าด้วยกัน
    all_data = []
    csv_files = [f for f in os.listdir(dir_path) if f.endswith('.csv')]

    for f in csv_files:
        try:
            temp_df = pd.read_csv(os.path.join(dir_path, f))
            all_data.append(temp_df[[station_col, target_col]])
        except:
            continue

    if not all_data:
        print("ไม่พบข้อมูลในเดือนนี้")
        continue

    df_month = pd.concat(all_data)

    # 3. วิเคราะห์รายสถานี
    # นับจำนวน Record ที่มีจริง (ไม่นับค่าที่เป็น NaN)
    station_stats = df_month.groupby(station_col)[target_col].count().reset_index()
    station_stats.columns = [station_col, 'actual_count']

    # คำนวณส่วนต่าง
    station_stats['missing_count'] = days_in_month - station_stats['actual_count']
    # กรณีบางสถานีส่งข้อมูลซ้ำเกินจำนวนวัน ให้แก้ค่าติดลบเป็น 0
    station_stats.loc[station_stats['missing_count'] < 0, 'missing_count'] = 0

    station_stats['missing_percent'] = (station_stats['missing_count'] / days_in_month) * 100

    # 4. แสดงผล (เลือกเฉพาะ 10 สถานีแรกที่หายเยอะที่สุดมาโชว์ หรือโชว์ทั้งหมด)
    # เรียงลำดับจากสถานีที่ข้อมูลหายมากที่สุด
    station_stats = station_stats.sort_values(by='missing_percent', ascending=False)

    for index, row in station_stats.iterrows():
        print(f"{row[station_col]:<15} | {days_in_month:<10} | {int(row['missing_count']):<10} | {row['missing_percent']:>8.2f}%")

    # สรุปภาพรวมเดือน
    avg_missing = station_stats['missing_percent'].mean()
    print(f"\n>> ภาพรวมเดือน {month_str}: ข้อมูลหายเฉลี่ย {avg_missing:.2f}% จากทั้งหมด {len(station_stats)} สถานี")


📊 [ สรุปข้อมูลสูญหายรายสถานี: 2025-07 ]
Station ID      | Expected   | Missing    | Missing %
-------------------------------------------------------
TUKW            | 31         | 21         |    67.74%
BKHN            | 31         | 17         |    54.84%
VLGE31          | 31         | 16         |    51.61%
NPI002          | 31         | 16         |    51.61%
NKO001          | 31         | 14         |    45.16%
CGKM            | 31         | 9          |    29.03%
PRGB            | 31         | 9          |    29.03%
CPY012          | 31         | 8          |    25.81%
CLPK            | 31         | 8          |    25.81%
SLA002          | 31         | 6          |    19.35%
TGSK            | 31         | 4          |    12.90%
BSPN            | 31         | 4          |    12.90%
MEPK            | 31         | 3          |     9.68%
CTKN            | 31         | 3          |     9.68%
PICI            | 31         | 2          |     6.45%
BTHO            | 31         | 2       

In [26]:
import pandas as pd
import os
from calendar import monthrange

# --- ตั้งค่า Path ---
input_dir = '/content/drive/MyDrive/สอบ3เทอม2/Output - Filtered Daily Rain'
# สร้างโฟลเดอร์ใหม่สำหรับเก็บเฉพาะสถานีที่ผ่านเกณฑ์ (ข้อมูลหายไม่เกิน 20%)
final_output_dir = '/content/drive/MyDrive/สอบ3เทอม2/Output - Clean Final Data'

target_col = 'rainfall_value'
station_col = 'tele_station_oldcode'
threshold = 20.0  # กำหนดเกณฑ์เปอร์เซ็นต์ที่ยอมให้หายได้สูงสุด

if not os.path.exists(final_output_dir):
    os.makedirs(final_output_dir)

months = ['2025-07', '2025-08', '2025-09', '2025-10', '2025-11', '2025-12']

for month_str in months:
    src_month_path = os.path.join(input_dir, month_str)
    dest_month_path = os.path.join(final_output_dir, month_str)

    if not os.path.isdir(src_month_path):
        continue

    if not os.path.exists(dest_month_path):
        os.makedirs(dest_month_path)

    print(f"\n🔍 กำลังตรวจสอบเดือน: {month_str}")

    # 1. คำนวณหาจำนวนวันที่ควรมีในเดือนนั้น
    year, month = map(int, month_str.split('-'))
    days_in_month = monthrange(year, month)[1]

    # 2. อ่านไฟล์ทั้งหมดเพื่อระบุสถานีที่ต้องตัดออก
    csv_files = [f for f in os.listdir(src_month_path) if f.endswith('.csv')]
    temp_list = []
    for f in csv_files:
        temp_list.append(pd.read_csv(os.path.join(src_month_path, f)))

    df_all_month = pd.concat(temp_list)

    # คำนวณเปอร์เซ็นต์หายรายสถานี
    station_stats = df_all_month.groupby(station_col)[target_col].count().reset_index()
    station_stats['missing_percent'] = (1 - (station_stats[target_col] / days_in_month)) * 100

    # คัดเลือกรายชื่อสถานีที่ผ่านเกณฑ์ (Missing <= 20%)
    valid_stations = station_stats[station_stats['missing_percent'] <= threshold][station_col].tolist()
    dropped_stations = station_stats[station_stats['missing_percent'] > threshold][station_col].tolist()

    print(f"   ✅ สถานีที่ผ่านเกณฑ์: {len(valid_stations)} สถานี")
    print(f"   ❌ สถานีที่ถูกตัดออก (>20%): {len(dropped_stations)} สถานี")
    if dropped_stations:
        print(f"      รายชื่อ: {', '.join(dropped_stations)}")

    # 3. บันทึกไฟล์ใหม่เฉพาะสถานีที่ผ่านเกณฑ์
    for f in csv_files:
        df_daily = pd.read_csv(os.path.join(src_month_path, f))
        # กรองเฉพาะสถานีที่อยู่ในลิสต์ valid_stations
        df_clean = df_daily[df_daily[station_col].isin(valid_stations)]

        # บันทึกลงในที่อยู่ใหม่
        df_clean.to_csv(os.path.join(dest_month_path, f), index=False)

print("\n✨ ประมวลผลเสร็จสิ้น! ไฟล์ที่กรองแล้วอยู่ที่โฟลเดอร์: Output - Clean Final Data")


🔍 กำลังตรวจสอบเดือน: 2025-07
   ✅ สถานีที่ผ่านเกณฑ์: 360 สถานี
   ❌ สถานีที่ถูกตัดออก (>20%): 9 สถานี
      รายชื่อ: BKHN, CGKM, CLPK, CPY012, NKO001, NPI002, PRGB, TUKW, VLGE31

🔍 กำลังตรวจสอบเดือน: 2025-08
   ✅ สถานีที่ผ่านเกณฑ์: 351 สถานี
   ❌ สถานีที่ถูกตัดออก (>20%): 12 สถานี
      รายชื่อ: BHRA, CHI010, CHI013, KKTO, MMMO, NEAK, NGPI, PSLT, SGKW, SKM004, VLGE31, WAN005

🔍 กำลังตรวจสอบเดือน: 2025-09
   ✅ สถานีที่ผ่านเกณฑ์: 343 สถานี
   ❌ สถานีที่ถูกตัดออก (>20%): 21 สถานี
      รายชื่อ: BHRA, BKHN, BNLG, BSPN, CEHM, CGKM, CHI013, CPY010, CTKN, DCCT, DGSL, KKTO, MEWG, MMMO, NBON, NGCG, NGPI, SKE001, STH011, VLGE31, WNKH

🔍 กำลังตรวจสอบเดือน: 2025-10
   ✅ สถานีที่ผ่านเกณฑ์: 347 สถานี
   ❌ สถานีที่ถูกตัดออก (>20%): 19 สถานี
      รายชื่อ: BKHN, BNGR, BNLG, BPTG, BSKJ, BSPN, CHI013, MUN014, NAPK, NASW, NEAK, NKSG, NMPR, NNPT, SKM004, SLA002, VLGE01, WNKH, YOM001

🔍 กำลังตรวจสอบเดือน: 2025-11
   ✅ สถานีที่ผ่านเกณฑ์: 365 สถานี
   ❌ สถานีที่ถูกตัดออก (>20%): 12 สถานี
      รายชื่อ: BKK0

#ขั้นตอนที่3 เติมข้อมูลด้วย Mediam ของ Day of Year จาก historical

In [30]:
import pandas as pd
import os

# --- 1. กำหนดตำแหน่งไฟล์และโฟลเดอร์ ---
historical_file = '/content/drive/MyDrive/สอบ3เทอม2/Daily rain - (Statistics 2012-2024)/hii_daily_rain.csv'
clean_data_dir = '/content/drive/MyDrive/สอบ3เทอม2/Output - Clean Final Data'
imputed_output_dir = '/content/drive/MyDrive/สอบ3เทอม2/Output - Final Imputed Data'

# ชื่อคอลัมน์ในไฟล์ประวัติศาสตร์ (Historical)
hist_station_col = 'station_name'
hist_data_col = 'data'
hist_date_col = 'date' # *** รบกวนตรวจสอบชื่อคอลัมน์วันที่ในไฟล์นี้อีกครั้ง (เช่น date หรือ rainfall_datetime) ***

# ชื่อคอลัมน์ในไฟล์ปัจจุบัน (2025)
curr_station_col = 'tele_station_oldcode'
curr_data_col = 'rainfall_value'
curr_date_col = 'rainfall_datetime'

if not os.path.exists(imputed_output_dir):
    os.makedirs(imputed_output_dir)

# --- 2. คำนวณค่า Median จากข้อมูลประวัติศาสตร์ ---
print("⏳ กำลังโหลดข้อมูลประวัติศาสตร์และคำนวณค่า Median...")
try:
    # อ่านเฉพาะคอลัมน์ที่จำเป็นเพื่อประหยัดหน่วยความจำ
    df_hist = pd.read_csv(historical_file)

    # ตรวจสอบชื่อคอลัมน์วันที่ (ถ้าไม่ใช่ 'date' ให้เปลี่ยนชื่อในตัวแปร hist_date_col)
    if hist_date_col not in df_hist.columns:
        # ลองหาคอลัมน์ที่มีคำว่า 'date' หรือ 'time'
        possible_date_cols = [c for c in df_hist.columns if 'date' in c.lower() or 'time' in c.lower()]
        if possible_date_cols:
            hist_date_col = possible_date_cols[0]
            print(f"🔎 พบคอลัมน์วันที่ในไฟล์ประวัติศาสตร์ชื่อ: {hist_date_col}")

    df_hist[hist_date_col] = pd.to_datetime(df_hist[hist_date_col])
    df_hist['day_of_year'] = df_hist[hist_date_col].dt.dayofyear

    # คำนวณ Median แยกตามชื่อสถานีและวันในรอบปี
    median_map = df_hist.groupby([hist_station_col, 'day_of_year'])[hist_data_col].median().reset_index()
    median_map.columns = [hist_station_col, 'day_of_year', 'median_val']
    print(f"✅ คำนวณ Median สำเร็จ (พบข้อมูล {len(median_map)} รายการ)")
except Exception as e:
    print(f"❌ เกิดข้อผิดพลาดในการประมวลผลไฟล์ประวัติศาสตร์: {e}")
    median_map = None

# --- 3. เริ่มกระบวนการเติมข้อมูล (Imputation) ---
if median_map is not None:
    months = ['2025-07', '2025-08', '2025-09', '2025-10', '2025-11', '2025-12']

    for month_str in months:
        src_path = os.path.join(clean_data_dir, month_str)
        dest_path = os.path.join(imputed_output_dir, month_str)

        if not os.path.isdir(src_path): continue
        if not os.path.exists(dest_path): os.makedirs(dest_path)

        print(f"🔹 กำลังเติมข้อมูลเดือน: {month_str}")
        csv_files = [f for f in os.listdir(src_path) if f.endswith('.csv')]

        for f in csv_files:
            df = pd.read_csv(os.path.join(src_path, f))
            df[curr_date_col] = pd.to_datetime(df[curr_date_col])
            df['day_of_year'] = df[curr_date_col].dt.dayofyear
            df[curr_station_col] = df[curr_station_col].astype(str)

            # นับค่าว่างก่อนเติม
            if df[curr_data_col].isna().any():
                # รวมข้อมูลโดยจับคู่ curr_station_col (2025) เข้ากับ hist_station_col (Historical)
                df = df.merge(median_map,
                              left_on=[curr_station_col, 'day_of_year'],
                              right_on=[hist_station_col, 'day_of_year'],
                              how='left')

                # เติมค่าว่างด้วย Median
                df[curr_data_col] = df[curr_data_col].fillna(df['median_val'])

                # หากยังว่าง (กรณีประวัติศาสตร์ไม่มีข้อมูลวันนั้น) ให้เติม 0
                df[curr_data_col] = df[curr_data_col].fillna(0)

                # ลบคอลัมน์ส่วนเกิน
                df.drop(columns=[hist_station_col, 'median_val', 'day_of_year'], inplace=True)
            else:
                df.drop(columns=['day_of_year'], inplace=True)

            # บันทึกไฟล์ที่สมบูรณ์แล้ว
            df.to_csv(os.path.join(dest_path, f), index=False)

    print(f"\n✨ เรียบร้อย! ข้อมูลที่สมบูรณ์ถูกเก็บไว้ที่: {imputed_output_dir}")

⏳ กำลังโหลดข้อมูลประวัติศาสตร์และคำนวณค่า Median...
🔎 พบคอลัมน์วันที่ในไฟล์ประวัติศาสตร์ชื่อ: measure_datetime
✅ คำนวณ Median สำเร็จ (พบข้อมูล 487922 รายการ)
🔹 กำลังเติมข้อมูลเดือน: 2025-07
🔹 กำลังเติมข้อมูลเดือน: 2025-08
🔹 กำลังเติมข้อมูลเดือน: 2025-09
🔹 กำลังเติมข้อมูลเดือน: 2025-10
🔹 กำลังเติมข้อมูลเดือน: 2025-11
🔹 กำลังเติมข้อมูลเดือน: 2025-12

✨ เรียบร้อย! ข้อมูลที่สมบูรณ์ถูกเก็บไว้ที่: /content/drive/MyDrive/สอบ3เทอม2/Output - Final Imputed Data


#ขั้นตอนที่4 รวมปริมาณฝนแต่ละวันเป็นผลรายเดือน

In [31]:
import pandas as pd
import os

# --- 1. กำหนดตำแหน่งโฟลเดอร์ ---
# โฟลเดอร์ที่เก็บข้อมูลซึ่งเติมค่า Median เรียบร้อยแล้วจากขั้นตอนก่อนหน้า
imputed_data_dir = '/content/drive/MyDrive/สอบ3เทอม2/Output - Final Imputed Data'
output_final_file = '/content/drive/MyDrive/สอบ3เทอม2/Y_monthly_rain_HII_final_summary.csv'

# รายชื่อเดือนที่ต้องการประมวลผล
months_list = ['2025-07', '2025-08', '2025-09', '2025-10', '2025-11', '2025-12']

final_summary = []

print("⏳ กำลังเริ่มคำนวณผลรวมรายเดือน...")

for folder_name in months_list:
    folder_path = os.path.join(imputed_data_dir, folder_name)

    if not os.path.isdir(folder_path):
        print(f"⚠️ ไม่พบโฟลเดอร์: {folder_name} (ข้าม...)")
        continue

    # แยกปีและเดือนออกมาจากชื่อโฟลเดอร์
    year_val = int(folder_name.split('-')[0])
    month_val = int(folder_name.split('-')[1])

    print(f"🔹 กำลังประมวลผลเดือน: {folder_name}")

    # ดึงไฟล์ CSV ทั้งหมดในโฟลเดอร์ของเดือนนั้นๆ
    csv_files = [f for f in os.listdir(folder_path) if f.endswith('.csv')]

    month_data_list = []
    for f in csv_files:
        try:
            temp_df = pd.read_csv(os.path.join(folder_path, f))
            # เลือกเฉพาะคอลัมน์ที่จำเป็น (รหัสสถานี และ ค่าฝน)
            if 'tele_station_oldcode' in temp_df.columns and 'rainfall_value' in temp_df.columns:
                month_data_list.append(temp_df[['tele_station_oldcode', 'rainfall_value']])
        except Exception as e:
            print(f"  ❌ ไม่สามารถอ่านไฟล์ {f} ได้: {e}")

    if month_data_list:
        # รวมข้อมูลทุกวันของเดือนนั้นเข้าด้วยกัน
        df_month = pd.concat(month_data_list)

        # คำนวณผลรวมรายเดือนแยกตามสถานี
        monthly_sum = df_month.groupby('tele_station_oldcode')['rainfall_value'].sum().reset_index()

        # ปรับรูปแบบคอลัมน์ให้ตรงตามไฟล์ตัวอย่าง
        monthly_sum['month'] = month_val
        monthly_sum['year'] = year_val
        monthly_sum.rename(columns={'tele_station_oldcode': 'Node', 'rainfall_value': 'rain'}, inplace=True)

        final_summary.append(monthly_sum)

# --- 2. รวมข้อมูลทุกเดือนและบันทึกผล ---
if final_summary:
    df_final = pd.concat(final_summary)

    # เรียงลำดับตาม Node และ Month เพื่อความสวยงามเหมือนไฟล์ตัวอย่าง
    df_final = df_final.sort_values(by=['Node', 'year', 'month']).reset_index(drop=True)

    # บันทึกไฟล์ CSV
    df_final.to_csv(output_final_file, index=False)

    print(f"\n✨ เสร็จสิ้นการประมวลผลทั้งหมด!")
    print(f"✅ บันทึกไฟล์สรุปรายเดือนที่: {output_final_file}")

    # แสดงตัวอย่างข้อมูล 5 แถวแรก
    print("\n--- ตัวอย่างข้อมูลผลลัพธ์ ---")
    print(df_final.head())
else:
    print("❌ ไม่พบข้อมูลสำหรับสร้างไฟล์สรุป")

⏳ กำลังเริ่มคำนวณผลรวมรายเดือน...
🔹 กำลังประมวลผลเดือน: 2025-07
🔹 กำลังประมวลผลเดือน: 2025-08
🔹 กำลังประมวลผลเดือน: 2025-09
🔹 กำลังประมวลผลเดือน: 2025-10
🔹 กำลังประมวลผลเดือน: 2025-11
🔹 กำลังประมวลผลเดือน: 2025-12

✨ เสร็จสิ้นการประมวลผลทั้งหมด!
✅ บันทึกไฟล์สรุปรายเดือนที่: /content/drive/MyDrive/สอบ3เทอม2/Y_monthly_rain_HII_final_summary.csv

--- ตัวอย่างข้อมูลผลลัพธ์ ---
   Node   rain  month  year
0  ACRU   83.0      7  2025
1  ACRU  164.8      8  2025
2  ACRU  248.6      9  2025
3  ACRU   48.4     10  2025
4  ACRU   52.2     11  2025
